src/main/java/org/alfresco/rest/api/impl/TagsImpl.java

```diff
@@ -125,20 +126,40 @@ public void deleteTag(String nodeId, String tagId)
     	taggingService.removeTag(nodeRef, tagValue);
     }
 
-    public CollectionWithPagingInfo<Tag> getTags(StoreRef storeRef, Paging paging)
-    {
-    	PagingResults<Pair<NodeRef, String>> results = taggingService.getTags(storeRef, Util.getPagingRequest(paging));
-    	Integer totalItems = results.getTotalResultCount().getFirst();
-    	List<Pair<NodeRef, String>> page = results.getPage();
-    	List<Tag> tags = new ArrayList<Tag>(page.size());
-    	for(Pair<NodeRef, String> pair : page)
-    	{
-    		tags.add(new Tag(pair.getFirst(), pair.getSecond()));
-    	}
+    public CollectionWithPagingInfo<Tag> getTags(StoreRef storeRef, Parameters params)
+    {
+        Paging paging = params.getPaging();
+        PagingResults<Pair<NodeRef, String>> results = taggingService.getTags(storeRef, Util.getPagingRequest(paging));
+        taggingService.getPagedTags(storeRef, 0, paging.getMaxItems());
+        Integer totalItems = results.getTotalResultCount().getFirst();
+        List<Pair<NodeRef, String>> page = results.getPage();
+        List<Tag> tags = new ArrayList<Tag>(page.size());
+        List<Pair<String, Integer>> tagsByCount = null;
+
+        if (params.getInclude().contains(PARAM_INCLUDE_COUNT))
+        {
+            tagsByCount = taggingService.findTaggedNodesAndCountByTagName(storeRef);
```

Comment:
This should be considered to be a map or added to a map, else in case we have a lot of tags used in the system the next search will have a very high complexity.